substrate/runtime/primitives/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate num_traits;
+extern crate substrate_runtime_std as rstd;
+#[cfg(not(feature = "std"))] extern crate substrate_runtime_io as runtime_io;
+extern crate substrate_codec as codec;
+extern crate substrate_primitives;
+
+use rstd::prelude::*;
+use codec::Slicable;
+pub use num_traits::identities::{Zero, One};
+use rstd::ops::{Add, Sub, Rem, AddAssign, SubAssign, RemAssign};
+
+pub trait HasPublicAux {
+	type PublicAux;
+}
+
+pub trait Hashing {
+	type Output;
+	fn hash_of<S: Slicable>(s: &S) -> Self::Output;
+	fn enumerated_trie_root(items: &Vec<&[u8]>) -> Self::Output;
+	fn storage_root() -> Self::Output;
+}
+
+pub trait RefInto<T> {
+	fn into(&self) -> &T;
+}
+impl<T> RefInto<T> for T {
+	fn into(&self) -> &T { &self }
+}
+
+pub trait SimpleArithmetic:
+	Zero + One +
+	Add<Self, Output = Self> + AddAssign<Self> +
+	Sub<Self, Output = Self> + SubAssign<Self> +
+	Rem<Self, Output = Self> + RemAssign<Self> +
+	PartialOrd<Self> + Ord
+{}
+impl<T:
+	Zero + One +
+	Add<Self, Output = Self> + AddAssign<Self> +
+	Sub<Self, Output = Self> + SubAssign<Self> +
+	Rem<Self, Output = Self> + RemAssign<Self> +
+	PartialOrd<Self> + Ord
+> SimpleArithmetic for T {}
+
+pub trait SimpleBitOps:
+	Sized +
+	rstd::ops::BitOr<Self, Output = Self> +
+	rstd::ops::BitAnd<Self, Output = Self>
+{}
+impl<T:
+	Sized +
+	rstd::ops::BitOr<Self, Output = Self> +
+	rstd::ops::BitAnd<Self, Output = Self>
+> SimpleBitOps for T {}
+
+/// Something that can be executed.
+pub trait Executable {
+	fn execute();
+}
+
+impl Executable for () {
+	fn execute() {}
+}
+impl<A: Executable, B: Executable> Executable for (A, B) {
+	fn execute() {
+		A::execute();
+		B::execute();
+	}
+}
+
+/// Something that acts like a `Digest` - it can have `Log`s `push`ed onto it and these `Log`s are
+/// each `Slicable`.
+pub trait Digesty {
+	type Item: Sized;
+	fn push(&mut self, item: Self::Item);
+}
+
+/// Something which fulfills the abstract idea of a Substrate header. It has types for a `Number`,
+/// a `Hash` and a `Digest`. It provides access to an `extrinsics_root`, `state_root` and
+/// `parent_hash`, as well as a `digest` and a block `number`.
+///
+/// You can also create a `new` one from those fields.
+pub trait Headery: Sized + Slicable {
+	type Number: Sized;
+	type Hash: Sized;
+	type Digest: Sized;
+	fn number(&self) -> &Self::Number;
+	fn extrinsics_root(&self) -> &Self::Hash;
+	fn state_root(&self) -> &Self::Hash;
+	fn parent_hash(&self) -> &Self::Hash;
+	fn digest(&self) -> &Self::Digest;
+	fn new(
+		number: Self::Number,
+		extrinsics_root: Self::Hash,
+		state_root: Self::Hash,
+		parent_hash: Self::Hash,
+		digest: Self::Digest
+	) -> Self;
+}
+
+/// Something which fulfills the abstract idea of a Substrate block. It has types for an
+/// `Extrinsic` piece of information as well as a `Header`.
+///
+/// You can get an iterator over each of the `extrinsics` and retrieve the `header`.
+pub trait Blocky {
```

Comment:
trait naming conventions from most idiomatic to least/worst:

Thing > Thingable > Thingy

Reply:
applies to just about everything in this file

Reply:
`Block` is already used as the concrete type. the abstract type is used a lot less than the concrete type. in any case, a punchy and consistent short form is better. And I really don't want to see `*Type` and `*Trait` suffixed everywhere.

Reply:
but this is a really, really bad name! We shouldn't have to import both anyway; the trait should only be used in situations which are generic -- this is why traits like `Read` are called `Read` and not `SomethingYouCanProbablyReadFromIGuess`.

> And I really don't want to see *Type and *Trait suffixed everywhere.

I agree, that would also be bad. I suggest that if we want to describe a trait that encompasses "Thing", we call the trait `Thing`, not `ThingTrait`, `Thingable`, or `Thingy`. The fact that it's a trait is already made clear by the fact that it's a trait.

Reply:
but then you get problematic phrasing like `impl Block for somewhere_else::Block { ... }`.

Reply:
I don't find that to be an issue. It's pretty clear that one is a trait and the other is not. In any case, it could be reimported as "blocky" for implementations, although I would still not recommend it.

justification in the standard library: there are `Iterator` implementations called `Iter`, `IterMut`, and `IntoIter` (which is the name often used for the iterator produced by an `IntoIterator` implementation).

Reply:
but surely that an argument in the opposite direction: standard library chose to mutate the trait name `*Iterator` for concrete types `*Iter`.

Reply:
the justification is for naming traits what they are. the `Iterator` and `IntoIterator` traits are named after what they do and haven't been mutated